runtime/tests/interpreter/dynamic_casting_test.go

```diff
@@ -3870,3 +3870,365 @@ func TestInterpretDynamicCastingReferenceCasting(t *testing.T) {
+
+		code := `
+			let x: Int? = 42
+			let y: Int? = x as? Int
+		`
+
+		inter := parseCheckAndInterpret(t, code)
+
+		AssertValuesEqual(
+			t,
+			inter,
+			interpreter.NewUnmeteredSomeValueNonCopying(interpreter.NewUnmeteredIntValueFromInt64(42)),
+			inter.Globals.Get("y").GetValue(inter),
+		)
+	})
+
+	t.Run("as!", func(t *testing.T) {
+		t.Parallel()
+
+		code := `
+			let x: Int? = 42
+			let y: Int = x as! Int
+		`
+
+		inter := parseCheckAndInterpret(t, code)
+
+		AssertValuesEqual(
+			t,
+			inter,
+			interpreter.NewUnmeteredIntValueFromInt64(42),
+			inter.Globals.Get("y").GetValue(inter),
+		)
+	})
+
+	t.Run("multi optional as!", func(t *testing.T) {
+		t.Parallel()
+
+		code := `
+			let x: Int??? = 42
+			let y: Int = x as! Int
+		`
+
+		inter := parseCheckAndInterpret(t, code)
+
+		AssertValuesEqual(
+			t,
+			inter,
+			interpreter.NewUnmeteredIntValueFromInt64(42),
+			inter.Globals.Get("y").GetValue(inter),
+		)
+	})
+
+	t.Run("multi optional as?", func(t *testing.T) {
+		t.Parallel()
+
+		code := `
+			let x: Int??? = 42
+			let y: Int? = x as? Int
+		`
+
+		inter := parseCheckAndInterpret(t, code)
+
+		AssertValuesEqual(
+			t,
+			inter,
+			interpreter.NewUnmeteredSomeValueNonCopying(interpreter.NewUnmeteredIntValueFromInt64(42)),
+			inter.Globals.Get("y").GetValue(inter),
+		)
+	})
+
+	t.Run("nil as?", func(t *testing.T) {
+		t.Parallel()
+
+		code := `
+			let x: Int? = nil
+			let y: Int? = x as? Int
+		`
+
+		inter := parseCheckAndInterpret(t, code)
+
+		AssertValuesEqual(
+			t,
+			inter,
+			interpreter.Nil,
+			inter.Globals.Get("y").GetValue(inter),
+		)
+	})
+
+	t.Run("nil as!", func(t *testing.T) {
+		t.Parallel()
+
+		code := `
+			fun test() {
+				let x: Int? = nil
+				let y: Int = x as! Int
+			}
+        `
+
+		inter := parseCheckAndInterpret(t, code)
+
+		_, err := inter.Invoke("test")
+		RequireError(t, err)
+
+		assert.ErrorAs(t, err, &interpreter.ForceCastTypeMismatchError{})
+	})
+
+	t.Run("string as!", func(t *testing.T) {
+		t.Parallel()
+
+		code := `
+			fun test(): String {
+				let hello: String???? = "hello"
+				let something: AnyStruct = hello 
+				return something as! String
+			}
+        `
+
+		inter := parseCheckAndInterpret(t, code)
+
+		result, err := inter.Invoke("test")
+		require.NoError(t, err)
+		require.Equal(t, interpreter.NewUnmeteredStringValue("hello"), result)
+	})
+
+	t.Run("return nested optional as?", func(t *testing.T) {
+		t.Parallel()
+
+		code := `
+			let x: Int??? = 42
+			let y: Int?? = x as? Int?
+		`
+
+		inter := parseCheckAndInterpret(t, code)
+
+		AssertValuesEqual(
+			t,
+			inter,
+			interpreter.NewUnmeteredSomeValueNonCopying(
+				interpreter.NewUnmeteredSomeValueNonCopying(
+					interpreter.NewUnmeteredIntValueFromInt64(42),
+				),
+			),
+			inter.Globals.Get("y").GetValue(inter),
+		)
+	})
+
+	t.Run("return nested optional as!", func(t *testing.T) {
+		t.Parallel()
+
+		code := `
+			let x: Int??? = 42
+			let y: Int?? = x as! Int??
```

Comment:
Can you please also add a test case for this scenario (`Int???` casted to `Int??`) with the `as?` operator?

Reply:
Added.